MdeModulePkg/Bus/Pci/PciBusDxe/PciDoe.c

```diff
@@ -0,0 +1,342 @@
+                           1,
+                           &DoeStatus
+                           );
+
+  DEBUG ((DEBUG_INFO, "Reading data from mailbox\n"));
+
+  Index = 0;
+  Data  = Buffer;
+
+  while (DoeStatus.Bits.DataObjectReady && (Index * 4) < *Count) {
+    // Read response
+    Status = PciIoDevice->PciIo.Pci.Read (
+                                      &PciIoDevice->PciIo,
+                                      EfiPciIoWidthUint32,
+                                      PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_READ_DATA_MAILBOX_OFFSET,
+                                      1,
+                                      &Data[Index]
+                                      );
+    // Clear the FIFO
+    Status |= PciIoDevice->PciIo.Pci.Write (
+                                       &PciIoDevice->PciIo,
+                                       EfiPciIoWidthUint32,
+                                       PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_WRITE_DATA_MAILBOX_OFFSET,
+                                       1,
+                                       &Data[Index]
+                                       );
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    Index++;
+  }
+
+  *Count = Index * 4;
+
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+PciIoDoeTypeCheck (
+  IN EDKII_PCI_DOE_PROTOCOL  *This,
+  IN     UINTN               VendorId,
+  IN     UINTN               Type
+  )
+{
+  EFI_STATUS                        Status;
+  PCI_IO_DEVICE                     *PciIoDevice;
+  PCI_EXPRESS_DOE_DISCOVERY_PACKET  Packet;
+  UINT8                             Index;
+  UINTN                             Count;
+  UINT16                            ResponseVid, ResponseType;
+  UINT32                            Response[2];
+
+  PciIoDevice = PCI_IO_DEVICE_FROM_PCI_DOE_THIS (This);
+
+  DoeWaitForNotBusy (PciIoDevice);
+
+  Packet.Packet.Header1 = (PCI_DOE_PROTOCOL_DISCOVERY << DOE_HEADER1_OFST_TYPE) | (PCI_VENDOR_ID_PCI_SIG << DOE_HEADER1_OFST_VID);
+  Packet.Packet.Header2 = 3 << DOE_HEADER2_OFST_LEN;
+
+  Index = 0;
+
+  do {
+    Packet.Packet.Dw0 = Index;
+
+    DEBUG ((DEBUG_INFO, "Sending DOE Discovery packet\n"));
+
+    // Send the DOE discovery packet
+    Count  = 3 * 4;
+    Status = PciIoDoeSend (This, Count, &Packet.Uint32);
+
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    // Set DOE Go bit
+    DoeSetGoBit (PciIoDevice);
+
+    // Wait for a response
+    DoeWaitForStatusDor (PciIoDevice);
+
+    // Read response
+    Count  = 2 * 4;
+    Status = PciIoDoeRecieve (This, &Count, &Response);
+
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    DEBUG ((DEBUG_INFO, "Response is: 0x%x/0x%x\n", Response[0], Response[1]));
+
+    Index        = Response[0] >> 24;
+    ResponseVid  = Response[0] & 0xFFFF;
+    ResponseType = (Response[0] >> 16) & 0xFFFF;
+

```

Comment:
This looks confusing and likely needs more comments.  Index is upper 8 bits of Response[0] and ResponseType is upper 16-bits of same Response[0].  This means the Index value is contained within ResponseType.  Is that intended?

Reply:
I added some comments and fixed `ResponseType`